keystorage/aws/src/test/java/tech/pegasys/signers/aws/AwsSecretsManagerTest.java

```
@@ -281,28 +311,49 @@ private String createSecret(final Tag tag, String secretValue)
     return testSecretName;
   }
 
-  private String createTestSecret(
-      final String tagKey, final String tagVal, final String secretValue) throws Exception {
-    final Tag testSecretTag = Tag.builder().key(tagKey).value(tagVal).build();
+  private boolean checkTestSecretExistsAndUpdateIfUnmatched(
+      final String testSecretName, final Tag tag, final String secretValue) {
     try {
-      return createSecret(testSecretTag, secretValue);
+      final GetSecretValueResponse getSecretValueResponse =
+          testSecretsManagerClient
+              .getSecretValue(GetSecretValueRequest.builder().secretId(testSecretName).build())
+              .get(30, TimeUnit.SECONDS);
+      final DescribeSecretResponse describeSecretResponse =
+          testSecretsManagerClient
+              .describeSecret(DescribeSecretRequest.builder().secretId(testSecretName).build())
+              .get(30, TimeUnit.SECONDS);
+      if (!getSecretValueResponse.secretString().equals(secretValue)
+          && describeSecretResponse.hasTags()
+          && !describeSecretResponse.tags().get(0).equals(tag)) {
+        updateSecret(testSecretName, secretValue, describeSecretResponse.tags().get(0), tag);
+      }
+      return true;
     } catch (Exception e) {
-      throw new Exception(e.getMessage());
+      return false;
```

If the secret exists but we throw an exception while checking and/or updating it then I think we should throw an exception here rather than return false. 

Returning false will result in a new CreateSecretRequest which would fail in this scenario.

---

Would it hurt to bottle up in that fashion? Don't think we care to debug when the secret fails to update, only when creating, listing, and fetching.

---

If the secret fails to update then this code would return false and therefore attempt to recreate the same secret, which would fail because it already exists.

I think we can only continue with the test if all attempted AWS calls (including updating) are successful, so in my mind it makes sense to always bubble up AWS exceptions.

---

in that case the test fails to run. We can then try again or manually inspect with the console

---

I guess not ideal on the first pipeline run